### PR TITLE
Refactor duplicate code, add missing method/encoding combinations

### PR DESCRIPTION
**Pull request by justinas**
Added a `PatchForm` method due to a use-case in cloud, and `PutForm` just for completeness' sake. Also took the liberty to factor out the common parts of form & json submission methods.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -166,11 +166,7 @@ func (c *Client) Endpoint(params ...string) string {
 	return fmt.Sprintf("%s/%s", c.addr, strings.Join(params, "/"))
 }
 
-// PostForm posts urlencoded form with values and returns the result
-//
-// c.PostForm(c.Endpoint("users"), url.Values{"name": []string{"John"}})
-//
-func (c *Client) PostForm(ctx context.Context, endpoint string, vals url.Values, files ...File) (*Response, error) {
+func (c *Client) doForm(ctx context.Context, method string, endpoint string, vals url.Values, files []File) (*Response, error) {
```
Comment (hatched) [inline client.go:169]: what about `submitForm` ?

File: client.go [hunk 2]
```diff
@@ -181,7 +177,7 @@ func (c *Client) PostForm(ctx context.Context, endpoint string, vals url.Values,
 
 	return c.RoundTrip(func() (*http.Response, error) {
 		if len(files) == 0 {
-			req, err := http.NewRequest(http.MethodPost, endpoint, strings.NewReader(vals.Encode()))
+			req, err := http.NewRequest(method, endpoint, strings.NewReader(vals.Encode()))
```
Comment (rosstimothy) [inline client.go:180]: nit: if we used `http.NewRequestContext` we could get rid of the need to use `req.WithContext`

File: client.go [hunk 3]
```diff
@@ -224,11 +220,31 @@ func (c *Client) PostForm(ctx context.Context, endpoint string, vals url.Values,
...
+
+// PatchForm patches urlencoded form with values and returns the result
+//
+// c.PatchForm(c.Endpoint("users"), url.Values{"name": []string{"John"}})
+//
+func (c *Client) PatchForm(ctx context.Context, endpoint string, vals url.Values, files ...File) (*Response, error) {
+	return c.doForm(ctx, http.MethodPatch, endpoint, vals, files)
+}
+
+func (c *Client) doJSON(ctx context.Context, method string, endpoint string, data interface{}) (*Response, error) {
```
Comment (hatched) [inline client.go:247]: `submitJSON` ?

**Review by hatched — APPROVED** [review]

File: client.go [hunk 4]
```diff
@@ -240,7 +256,7 @@ func (c *Client) PostJSON(ctx context.Context, endpoint string, data interface{}
 	tracer := c.newTracer()
 	return tracer.Done(c.RoundTrip(func() (*http.Response, error) {
 		data, err := json.Marshal(data)
-		req, err := http.NewRequest(http.MethodPost, endpoint, bytes.NewBuffer(data))
+		req, err := http.NewRequest(method, endpoint, bytes.NewBuffer(data))
```
Comment (rosstimothy) [inline client.go:259]: nit: if we used `http.NewRequestContext` we could get rid of the need to use `req.WithContext`

**Review by rosstimothy — APPROVED** [review]

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**